llvm/tools/llvm-objdump/llvm-objdump.cpp

```diff
@@ -1638,18 +1691,24 @@ disassembleObject(ObjectFile &Obj, const ObjectFile &DbgObj,
   LLVM_DEBUG(LVP.dump());
 
   std::unordered_map<uint64_t, BBAddrMap> AddrToBBAddrMap;
+  std::unordered_map<uint64_t, PGOAnalysisMap> AddrToPGOAnalysisMap;
   auto ReadBBAddrMap = [&](std::optional<unsigned> SectionIndex =
                                std::nullopt) {
     AddrToBBAddrMap.clear();
     if (const auto *Elf = dyn_cast<ELFObjectFileBase>(&Obj)) {
-      auto BBAddrMapsOrErr = Elf->readBBAddrMap(SectionIndex);
+      std::vector<PGOAnalysisMap> PGOAnalyses;
+      auto BBAddrMapsOrErr = Elf->readBBAddrMap(SectionIndex, &PGOAnalyses);
       if (!BBAddrMapsOrErr) {
         reportWarning(toString(BBAddrMapsOrErr.takeError()), Obj.getFileName());
         return;
       }
-      for (auto &FunctionBBAddrMap : *BBAddrMapsOrErr)
-        AddrToBBAddrMap.emplace(FunctionBBAddrMap.Addr,
-                                std::move(FunctionBBAddrMap));
+      for (auto &&[FunctionBBAddrMap, FunctionPGOAnalysis] :
```

Comment:
Why do we use `auto &&` here? We need non-const references. So `auto &` should be sufficient. right?

Reply:
It was suggested above by @red1bluelost. Not entirely sure what the reasoning was. Using a lvalue reference does create a temporary object here, but I don't think that should matter? The temporary object does prevent the creation of a non-const lvalue reference though, and I wouldn't really consider this to be a const use of the values  from the iterator.

Reply:
I don't think `auto &` and `auto &&` make a difference in the generated code for this case. Part of my reasoning (beyond it just being the habit I've built) for using `auto &&` is the fact that we move values out within the loop.

Reply:
Makes sense. Given `auto &` won't bind to the reference without a `const`, I think keeping it as `auto &&` makes sense given we're moving the values. @rlavaee what do you think?

Reply:
Thanks. I see what's happening now. Range over `zip_equal` returns a temporary tuple. So you can't take a non-const reference. `auto &&` is the right choice here. You can also use `zip_equal(*std::move(BBAddrMapsOrErr), std::move(PGOAnalyses))` to avoid unnecessary copies.

Reply:
Good point. Fixed in the latest iteration.